solr/contrib/s3-repository/src/java/org/apache/solr/s3/S3OutputStream.java

```diff
@@ -0,0 +1,260 @@
+  private MultipartUpload multiPartUpload;
+
+  public S3OutputStream(AmazonS3 s3Client, String key, String bucketName) {
+    this.s3Client = s3Client;
+    this.bucketName = bucketName;
+    this.key = key;
+    this.closed = false;
+    this.buffer = ByteBuffer.allocate(PART_SIZE);
+    this.progressListener = new ConnectProgressListener();
+    this.multiPartUpload = null;
+
+    if (log.isDebugEnabled()) {
+      log.debug("Created S3OutputStream for bucketName '{}' key '{}'", bucketName, key);
+    }
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    if (closed) {
+      throw new IOException("Stream closed");
+    }
+
+    buffer.put((byte) b);
+
+    // If the buffer is now full, push it to remote S3.
+    if (!buffer.hasRemaining()) {
+      uploadPart(false);
+    }
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (closed) {
+      throw new IOException("Stream closed");
+    }
+
+    if (outOfRange(off, b.length) || len < 0 || outOfRange(off + len, b.length)) {
+      throw new IndexOutOfBoundsException();
+    } else if (len == 0) {
+      return;
+    }
+
+    int currentOffset = off;
+    int lenRemaining = len;
+    while (buffer.remaining() < lenRemaining) {
+      int firstPart = buffer.remaining();
+      buffer.put(b, currentOffset, firstPart);
+      uploadPart(false);
+
+      currentOffset += firstPart;
+      lenRemaining -= firstPart;
+    }
+    if (lenRemaining > 0) {
+      buffer.put(b, currentOffset, lenRemaining);
+    }
+  }
+
+  private static boolean outOfRange(int off, int len) {
+    return off < 0 || off > len;
+  }
+
+  private void uploadPart(boolean isLastPart) throws IOException {
+
+    int size = buffer.position();
+
+    if (size == 0) {
+      // nothing to upload
+      return;
+    }
+
+    if (multiPartUpload == null) {
+      if (log.isDebugEnabled()) {
+        log.debug("New multi-part upload for bucketName '{}' key '{}'", bucketName, key);
+      }
+      multiPartUpload = newMultipartUpload();
+    }
+    try {
+      multiPartUpload.uploadPart(new ByteArrayInputStream(buffer.array()), size, isLastPart);
+    } catch (Exception e) {
+      if (multiPartUpload != null) {
+        multiPartUpload.abort();
+        if (log.isDebugEnabled()) {
+          log.debug("Multipart upload aborted for bucketName '{}' key '{}'.", bucketName, key);
+        }
+      }
+      throw new S3Exception("Part upload failed: ", e);
+    }
+
+    // reset the buffer for eventual next write operation
+    buffer.clear();
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (closed) {
+      throw new IOException("Stream closed");
+    }
+
+    // Flush is possible only if we have more data than the required part size
+    // If buffer size is lower than than, just skip
+    if (buffer.position() >= MIN_PART_SIZE) {
+      uploadPart(false);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
Is this thread safe? It doesn't look thread safe, but maybe it doesn't need to be.

Reply:
It definitely doesn't look threadsafe, but not sure whether it would need to be.